bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/AuthFilter.java

```diff
@@ -142,22 +188,17 @@ public void filter(ContainerRequestContext requestContext) throws IOException {
                         try {
                             String[] decodedCredentials = new String(Base64.getDecoder().decode(authParts[1]), "UTF-8")
                                     .split(":");
-                            if (decodedCredentials.length > 2) {
+                            if (decodedCredentials.length != 2) {
                                 throw new AuthenticationException("Invalid Basic authentication credential format");
                             }
-                            switch (decodedCredentials.length) {
-                                case 1:
```

Comment:
Why has this case been removed? How is it covered now? I cannot see it anymore.

Reply:
This was old legacy code that was not correct any more. It's trying to mix token based auth with basic auth.
It's either one or the other: If the header is basic auth it should be basic auth, if the header specifies token it should use token.
But specifying basic auth and trying to do token based auth is wrong and not necessary any more if basic auth works fine.

Reply:
How can I see that it is legacy code? Which class deals with the token header now? This isn't clear for me from the code. Anyhow, as it is unrelated to what you want to do in the PR, better leave the change out.

Reply:
This change will probably break existing setups as the current docs also state you can use tokens with basic authentication:

https://www.openhab.org/docs/configuration/restdocs.html#authentication

It maybe wrong it seems convenient. This way you can probably also easily test/use these tokens in your browser or in code by adding it in the URL as https://token@host/...

Reply:
Another way to make it less wrong but keep the convenient way, is by allowing such tokens to be passed in a parameter instead of a header. E.g. while integrating [Keycloak](https://www.keycloak.org/) in another project, I've seen it also allows for passing tokens with `https://host?access_token=...` . But it would still be a breaking change.

Reply:
I'd claim that the docs are simply wrong or at least an ugly workaround. Passing a bearer token [should be done this way](https://stackoverflow.com/a/25036577):
```
curl -H "Authorization: Bearer <TOKEN>" http://www.example.com
````
I'm happy to provide a PR to update the documentation, if you agree with it.

Reply:
I'm fine with dropping support for this when the docs are updated and it is mentioned in the release notes.

Reply:
> I've seen it also allows for passing tokens with https://host?access_token=...

Strongly advise against this. Long-lived tokens should never be part of any URL because of the risk they might end in access logs.

And **please** do not remove the "passing an API token as basic auth creds" since it helps in situations where passing it in another way is impossible. See the Node-Red example in the docs. It is very much not "legacy", GitHub (and others) is doing it albeit slightly differently (https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token#using-a-token-on-the-command-line) and, by the way, they deprecated password authorization to the API (https://docs.github.com/en/rest/overview/other-authentication-methods#via-username-and-password), which was the right thing to do.

Reply:
I agree using tokens is the way forward so it is quite useful to also have a way to use them with applications that cannot easily add add the right header. It's not an improvement if we would again force our users to create users with passwords and enable basic authentication to use the API with such applications.

Reply:
@ghys I'd still say that the docs should be changed to clearly say that tokens are supposed to be passed as headers and to only keep the Basic Auth workaround as a last resort.
But also there, I don't like that it is `<token>:` for `<username>:<password>`. It would be much cleaner to require it the same way as Github, i.e. as `<username>:<token>`. Our tokens are always associated to users, aren't they?

Reply:
> @ghys I'd still say that the docs should be changed to clearly say that tokens are supposed to be passed as headers and to only keep the "token passing as Basic Auth" workaround as a last resort.

Absolutely. I overlooked that part.

> But also there, I don't like that it is <token>: for <username>:<password>. It would be much cleaner to require it the same way as Github, i.e. as <username>:<token>. Our tokens are always associated to users, aren't they?

Agreed as well, this very point was discussed in the forums a while ago but I can't seem to find it anymore - in hindsight this should have been done like this. Maybe keep the old way too (pass the API token as username) at least temporarily for retrocompatibility's sake? So the recommended way to access the API would be, in order:
1. `Authorization: Bearer <JWT token>` (i.e. prefer OAuth is possible - web clients, native clients able to spawn a webview for OAuth authorization...)
2. `Authorization: Bearer <API token>`
3. `X-OPENHAB-TOKEN: <JWT token or API token>` (if the Authorization header is already taken for authentication to a reverse proxy)
4. `Authorization: Basic base64(<username>:<API token>)` (alternative to already implemented 5. below; if the client only supports Basic auth)
5. `Authorization: Basic base64(<API token>:)` (API token as username and blank password, the currently implemented option; maybe don't document it so usage will drop)
6. `Authorization: Basic base64(<username>:<password>`) **(DISCOURAGED and needs the option activated in Settings > API Security)**.

Note that the "Allow Basic authorization" option should perhaps be renamed to "Allow password-based Basic authentication", if you don't want to remove it, since it could be assumed that 4./5. require this option to work when they don't. Maybe clearly indicate that it's not recommended to use that option, since we've already seen what it does with improperly configured proxies forwarding their own Basic auth to openHAB - in some cases it will try to validate the proxy's password and this will slow the UI to a crawl.

Reply:
> Maybe clearly indicate that it's not recommended to use that option, since we've already seen what it does with improperly configured proxies forwarding their own Basic auth to openHAB - in some cases it will try to validate the proxy's password and this will slow the UI to a crawl.

This PR fixes that issue.

> by the way, they deprecated password authorization to the API (https://docs.github.com/en/rest/overview/other-authentication-methods#via-username-and-password), which was the right thing to do.

A web facing service with 40 million users should definitely have increased security over something that runs locally in your network because it has a much higher attack surface. But since openhab doesn't enforce ssl certifactes what we're doing is all just smoke and mirrors anyway. Basic Auth is just as secure as long living tokens except for the password leakage and there have been no constructive arguments that prove otherwise.

> And please do not remove the "passing an API token as basic auth creds" since it helps in situations where passing it in another way is impossible.

Please do, this is nothing more than a very bad hack. If the header specifies basic auth, that's what it should be!
What's next - a POST request that deletes items, a GET request that sets values (the one from classic UI was removed with this exact argument). If we do this, then why use standards at all?

@ghys 
Why do you get to make the decision about API security and not the user?
If I have openhab running as a microservice with an authenticating reverse proxy with an ssl connection, why do I have to pass credentials to openhab which then validates it again. Why can't I just pass a user and oh will accept it without further performance penalty.
I'm the last person to say we don't need to secure things but it has to be in such a way that it
- works for many use cases
- works with good performance
- can be configured by the user


@kaikreuzer @ghys @wborn 
There is definitely a need to interact with openhab in a very fast way from other programs.
Maybe the REST API is not the right tool for that after all.
HomeAssistant and NodeRed also offer websockets as an in/output. Authentication is done **once** when the websocket connects and thus the performance impact is negligible (That also the reason there was no "drama" when HomeAssistant switched to token based auth because it did not break anything performance wise).
Maybe that way is worth investigating?

Reply:
> Why do you get to make the decision about API security and not the user?
If I have openhab running as a microservice with an authenticating reverse proxy with an ssl connection, why do I have to pass credentials to openhab

As said before, if you want to run openhab as a microservice and roll your own auth, you're not in the "standard" supported use case and should modify the Karaf config or fork the distro to disable the built-in auth bundle. The UI was fixed to work in these cases. If you want the built-in auth, and most people do, then you play by its rules and avoid basic auth.

Reply:
> if you don't want to remove it, since it could be assumed that 4./5. require this option to work when they don't. Maybe clearly indicate that it's not recommended to use that option, since we've already seen what it does with improperly configured proxies forwarding their own Basic auth to openHAB

I think we should also mention it is deprecated and will be removed eventually so users stay clear of it and migrate to tokens. With 4. and 5. in place what valid reasons are there to cling to 6.? It's a bad idea to recycle the same user credentials in an application using the API.

Reply:
So to recap:

- You've come out of the blue [advocating](https://github.com/openhab/openhab-core/issues/1699) for the removal of my spare-time work on OAuth2, instead of finding solutions, because it didn't fit YOUR use case
- I recognized it was not ideal (despite its benefits) at the time and quickly added first basic auth (https://github.com/openhab/openhab-core/pull/1713) (which I immediately thought was a mistake) and then one week later API tokens (https://github.com/openhab/openhab-core/pull/1735) mostly (but of course not only) for YOU
- I researched how to disable auth from an existing instance (https://github.com/openhab/openhab-core/issues/2060#issuecomment-757392996), commented on why it shouldn't be easy to do especially remotely, and also fixed the UI (https://github.com/openhab/openhab-webui/pull/812) to allow it to work
- I tried my best to explain to you why passwords were deliberately slow to validate and tokens were not (https://github.com/openhab/openhab-core/pull/2030#issuecomment-757576201); I made the point that tokens were a better fit for API usage and using them is faster and keeps the password verifications to a minimum (not to mention storage/transport); you didn't even acknowledge that point and quoted me out of context (https://github.com/openhab/openhab-core/pull/2030#issuecomment-757651574) - I hope not deliberately
- you repeatedly ignored advice to even try using tokens and try benchmarking if it solves your perf issues (https://github.com/openhab/openhab-core/issues/1899#issuecomment-739996254, https://github.com/openhab/openhab-core/issues/1899#issuecomment-768097895)
- you implemented this PR anyways against multiple reservations and silently removed a documented feature that allows people to still enjoy the benefits of using a token when their clients don't allow it because it's "a bad hack" (that you will have to do soon if you want to continue Git using GitHub)
- you nagged the maintainers until you got your review (and refactoring)

At that point, let me ask, who's selfish, and who's constructive, and who has the attitude? You tell me.

No need to waste your keystrokes to answer. I'm done here, and I'm recusing myself from this conversation.

Reply:
It's funny that you feel the same way as I do, but with switched personas:

- In this issue I've pointed out that auth will make problems and just asked for a config switch to disable it. And here we are, arguing back and forth still not having a feasible solution.
- I was glad that you jumped in and created the basic auth option
- I asked for help to disable auth and I got an answer that broke the UI so of course I felt betrayed, especially since my hint that it breaks the UI has been [left unanswerd](https://github.com/openhab/openhab-core/issues/2060#issuecomment-757419590). I haven't gotten the news that you fixed this issue and I was under the impression it'll stay that way. It would have been nice if you would have linked the pr to fix the issue.
- I understand why passwords are deliberately slow to validate and even acknowledged that this is a good thing in my [comment](https://github.com/openhab/openhab-core/pull/2030#issuecomment-757651574). However in my comment I also describe that an attacker gains no new information from sending the same auth twice so subsequent requests should be quick to validate. That point and the differentiation between passwords in memory and password storage/transport has been (I hope not deliberately) ignored by you.
- I couldn't switch to token based auth because I'm depending on sse-libraries that don't allow setting custom headers. I know that's bad, but there aren't many sse libraries in python so I'll need to jump in and help refactoring but that's something that still has to be done
- When it was clear that there would be no quick option that'll fix my application I decided to jump in and provide an implementation in java myself which is not the language of my choice. The implementations also happens to fix many of the reverse proxy issues apparent in the forum
- True

I tried my best to make a simple solution work but it sure felt that you tried every time to shoot it down.

Reply:
> Why remove it - you loose nothing if it's opt in.

We remove unneccessary features because it takes time to maintain, document and the extra configuration options confuse users. I.e. the same reasons as why the compatibility layer and various UIs were removed in OH 3.

> I couldn't switch to token based auth because I'm depending on sse-libraries that don't allow setting custom headers. I know that's bad, but there aren't many sse libraries in python so I'll need to jump in and help refactoring but that's something that still has to be done

Did you try using a token as username and an empty password with the library? That should result in the same kind of Basic authentication header that should already work since 4. is already implemented. If it doesn't work because of the empty password, it should work when 5. is implemented.

Reply:
Hey guys, please calm down again. I don't understand why the discussion suddenly escalates like this, I had the hope that we were almost done and all that is left is to add a "Github-like token passing with `Authorization` header" for tools that can't deal with adding proper HTTP Bearer headers. I'm volunteering to (re-)implement this according to @ghys [summary above](https://github.com/openhab/openhab-core/pull/2101#discussion_r598359939). I think with that we should all be ok to accept this PR and everyone should be served.

Reply:
> Hey guys, please calm down again. I don't understand why the discussion suddenly escalates like this, 

Sorry Kai, I know I overreacted, with the hundreds of hours I pour into this project out of "love", really (I don't even use the auth layer, I just want it to be "right"), being accused of being "selfish"? This was just the last straw.

Reply:
@ghys:
I'm sorry I called your behavior selfish and about the too rough tone. My intention was never to insult you.
I've - like you - poured hundreds of hours into openhab in the form of a 3rd party rule engine and all I could see was it breaking and my countless efforts being all for nothing.

Reply:
> If it doesn't work because of the empty password, it should work when 5. is implemented.

Yep - I'll make sure to check again once 5. is implemented

@kaikreuzer 
Maybe we hash the [whole auth header](https://github.com/openhab/openhab-core/blob/ed42fd0a82762fcc1944a4d4e2d253ae37ad508e/bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/AuthFilter.java#L228) and use it as a key for the cache? That way everything gets a speedup independent of the authentication method.